File: android/codec/src/main/kotlin/com/farosos/codec/CaseASignature.kt
package com.farosos.codec

import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer

/**
 * Autocertificado de Caso A (`spec/packet-format.md`, sección
 * `FRAGMENTO_FIRMA`, decisión 18): el dispositivo firma su propia clave
 * pública Ed25519 — prueba posesión de la identidad, no autentica el
 * contenido de ningún beacon individual (límite aceptado, ver spec).
 * Vectores de prueba en `spec/test-vectors.json`, clave
 * `fragmento_firma.identity`. Ed25519 vía BouncyCastle — mismo motivo que en
 * `:app/DeviceIdentity` (#41): `java.security` no lo soporta de forma
 * confiable a minSdk 26.
 */
object CaseASignature {
    /** Tamaños fijos de Ed25519 — fuente única para quien necesite partir o validar un payload `pubkey || firma` (`SignatureFragmenter`). */
    const val PUBLIC_KEY_LENGTH: Int = 32
    const val SIGNATURE_LENGTH: Int = 64

    /** `firma = Ed25519_Sign(privkey, pubkey)`. */
    fun sign(privateKey: Ed25519PrivateKeyParameters): ByteArray {
        val publicKey = privateKey.generatePublicKey().encoded
        val signer = Ed25519Signer()
        signer.init(true, privateKey)
        signer.update(publicKey, 0, publicKey.size)
        return signer.generateSignature()
    }

    /**
     * Verifica que [signature] sea un autocertificado válido de [publicKey]
     * (es decir, [signature] firma [publicKey] bajo esa misma clave).
     */
    fun verify(publicKey: ByteArray, signature: ByteArray): Boolean {
        if (publicKey.size != PUBLIC_KEY_LENGTH) return false
        val verifier = Ed25519Signer()
        verifier.init(false, Ed25519PublicKeyParameters(publicKey, 0))
        verifier.update(publicKey, 0, publicKey.size)
        return runCatching { verifier.verifySignature(signature) }.getOrDefault(false)
    }
}

File: ios/Sources/PacketCodec/CaseASignature.swift
import CryptoKit
import Foundation

/// Autocertificado de Caso A (`spec/packet-format.md`, sección
/// `FRAGMENTO_FIRMA`, decisión 18): el dispositivo firma su propia clave
/// pública Ed25519 — prueba posesión de la identidad, no autentica el
/// contenido de ningún beacon individual (límite aceptado, ver spec).
/// Vectores de prueba en `spec/test-vectors.json`, clave
/// `fragmento_firma.identity`.
///
/// `CryptoKit.Curve25519.Signing` firma con un nonce aleatorio por firma
/// ("hedged", mitigación de fault-injection) — a diferencia de la firma
/// EdDSA determinística de RFC 8032 (usada por `@noble/curves` para generar
/// `signature_hex` en los vectores). Dos firmas de CryptoKit sobre el mismo
/// mensaje con la misma clave NO son iguales byte a byte entre sí, ni
/// contra el vector — ambas son igual de válidas. Por eso los tests de esta
/// clase verifican `verify(sign(...))`, nunca una igualdad byte a byte
/// contra `signature_hex`.
public enum CaseASignature {
    /// Tamaños fijos de Ed25519 — fuente única para quien necesite partir o
    /// validar un payload `pubkey || firma` (`SignatureFragmenter`).
    public static let publicKeyLength = 32
    public static let signatureLength = 64

    /// `firma = Ed25519_Sign(privkey, pubkey)`.
    public static func sign(privateKey: Curve25519.Signing.PrivateKey) -> Data {
        let publicKey = privateKey.publicKey.rawRepresentation
        // Ed25519 firma cualquier mensaje sin condiciones de error reales
        // para una clave válida — mismo patrón de `try!` que
        // `CaseBAuthentication` para operaciones garantizadas por CryptoKit.
        return try! privateKey.signature(for: publicKey)
    }

    /// Verifica que `signature` sea un autocertificado válido de `publicKey`
    /// (es decir, `signature` firma `publicKey` bajo esa misma clave).
    public static func verify(publicKey: Data, signature: Data) -> Bool {
        guard let key = try? Curve25519.Signing.PublicKey(rawRepresentation: publicKey) else { return false }
        return key.isValidSignature(signature, for: publicKey)
    }
}
